### ISSUE TITLE
Several URLs processed one by one

### ISSUE THREAD

**Issue by csaezl** [Norconex/collector-http#131]
I'm not sure if this can be done with the crawler. It's very simple. I need to process several URLs (say 100+ and more) each one with its own filter.
If I've understood your crawler behaviour, the following parameters:
```
      <startURLs>
          <url>URL-A</url>      
          <url>URL-B</url>      
      </startURLs>
...
      <referenceFilters>
        <filter class="$filterRegexRef" onMatch="include">Filter for URL-A</filter>       
        <filter class="$filterRegexRef" onMatch="include">Filter for URL-B</filter>       
      </referenceFilters>
```
can do the job but they might execute in parallel and every filter is applied to both URL, not only for the URL it is defined for.
The other way I know is to define different crawlers in the configuration file but they all might execute in parallel and I don't want it, I need a classical batch processing.
What is the proper way?

**Comment by essiembre** [comment]
The URLs will be processed all in "parallel" whether you list them all in the ``startURLs`` or in different crawlers.  I mean, it won't wait to finish with a site before moving to the next and different threads will work at it (unless you specify just 1 thread and 1 crawler, of course, and your delay can influence that too).  

That being said, if you want a filter to apply to a specific URL only, you have to have to build that filter regex to include that URL as well.  It may seem like overkill, but in a sense you would have to do that no matter what.  I mean, whether you associate a URL with a filter in a ``startURL`` tag or in a regex, you would still have to do that association, right? And overlap can occur regardless if your URLs do generate overlaps (a URL can link to another one you have defined, and apply its own filters on it).

So I think your best bet is to adjust your regex accordingly.  You can use variables to help as well. For instance, let's say you want to rely on the start of a URL to limit your filters, you can do something like this:
```xml
## These variables can be in an include file if you prefer:
#set($urlA = "http://www.urlA.com")
#set($urlB = "http://www.urlB.com")

      <startURLs>
          <url>${urlA}</url>      
          <url>${urlB}</url>      
      </startURLs>
...
      <referenceFilters>
        <filter class="$filterRegexRef" onMatch="include">${urlA}/.*/whateverPattern.*</filter>       
        <filter class="$filterRegexRef" onMatch="include">${urlB}/.*/whateverPattern.*</filter>       
      </referenceFilters>
```

That way you keep your regex relatively short.   If you have hundreds it can be tedious though.  Unless maybe you generate an include file somehow?  Keep in mind that the configs are read using Apache Velocity, so you can use other rendering logic as well, such as loop and conditions, right in the config.  Maybe this can help as well?

**Comment by csaezl** [comment]
Thank you for your  answer. As seen in other issues I have posted, I use the method you proposes: url included in the filter, I build them with a semi-automatic procedure. 

I'll apply one thread because I've experienced that with more that 1, the memory used by java machine seems to increase, even thought I have assigned some GB (about half the total amount) to it. I supposed the memory used by an URL would be recovered after being processed but it doesn't seem to happen.

From your answer I derive that an URL referenced from URL-A, that doesn't match URL-A filter but that does URL-B filter, would be processed as a valid URL.That is, there is no way of creating a relationship between an URL and a filter, that ensures that processing URL-A, only URLs that match URL-A filter will be processed. Right now the filters act as a pool for the URL pool.

One last thing. Using ```{urlA}``` in the filter, the regular expression has to deal with "." as ".", not as ```\.```. Does it work?.

**Comment by essiembre** [comment]
I am not sure I fully understand what you are after that the current filter mechanism does not provide.   All filters you define in a crawler apply to all URLs you have defined in that same crawler.  This issue is not inherent to the HTTP Collector, but to any crawler (Nutch, etc.).  I mean, let's say you have two crawlers (no matter which product) with different URLs and different filters.  If your crawler A accepts URLs that you have rejected in crawler B, Crawler A won't know about it an accept it.  So you will get that issue as well.   In other words, you always have to make sure you carefully delimiter in each crawler what you want to crawl, whether you have one or many URLs. By doing so, that's how you can limit crawling to just the URLs you want.

Can you give me a concrete example where you can't make current filtering work for you? I may be missing something.  Or do you have an alternate approach to propose?

As for the memory, it is the nature of Java to take memory as it needs to (up to the maximum heap size of the JVM), but release it only when it feels appropriate to do so (via its garbage collector).  So it is possible you see memory increase at every URLs and not come down for a long while, but that does not mean much in Java.  The important thing is that it does not exceed the maximum heap size as you would get an OutOfMemoryError.  Upon/if reaching close to the max heap, it will free memory up as needed. That's sometimes how the JVM optimizes things.  Running garbage collection has a cost so sometimes it only runs it when it really needs to.  

The HTTP collector has been used multi-threaded by most (and sometimes with more threads that I had imagined) without memory issues.  If you get an OutOfMemory error, this would be a bug that I encourage you to report (there have been some reported before... but they have all been fixed).

For the dots, that's just an example I gave you.  You may need to adapt as needed.  For most cases, I would be inclined to think that you are unlikely to find URLs matching a regex with the "wildcard" dots in the domain for something other.  Since domains always need dots in a URL, http://www.blah.com will not match other domains except if other dots are replaced to give weird domains, like http://wwwXblah.com.  Unlikely, but possible, so yes, you may have to adapt or use a different strategy.

**Comment by csaezl** [comment]
As for the filter, perhaps my explanation has complicated what I really need, that is nothing but to use an efficient way of batch processing URLs. I focused on filters because I thought they could be the solution for me.

I need to crawl every site in a independent way, so I guarantee that when crawling ```www.siteA.com```, the crawler only will process references belonging to ```www.siteA.com``` domain.

**Comment by csaezl** [comment]
As for the filter, perhaps my explanation has complicated what I really need, that is nothing but to use an efficient way of batch processing URLs. I focused on filters because I thought they could be the solution for me.

I need to crawl every site in a independent way, so I guarantee that when crawling ```www.siteA.com```, the crawler only processes references that belong to ```www.siteA.com``` domain. I do this with one configuration file that has one URL and one filter for that URL. That solves the need.

If I have 100 URL to crawl, I can build 100 configuration files, one for each URL. Or just one configuration file with 100 crawler definitions. Both strategies ensure that for each URL only references in the domain will be processed.

I simply thought that perhaps I could define one configuration file with 100 URL and 100 filters, that, when processing each URL just the related filter would be applied. But it doesn't work this way. I wonder if this processing method deserves to be implemented in HTTP Collector. As you can see, it's just a matter of batch processing.

**Comment by essiembre** [comment]
Ahhhh.... I see now.  It is not just facilitating some arbitrary filters you are after, but really a way to ensure crawler "stay on domain" for each URL.   I have seen such a feature implemented in another commercial crawler before.  I thought the flexible filters in HTTP Collector were making such a feature useless, but if only for convenience, maybe not.  

Would it make life significantly easier to see a new flag called "stayOnSite" (false|true) that you can optionally specify with each URL (or for all URLs in a seed file)?  What that would do is implicitly create a filter that forces extracting only URLs from a page that matches the URL domain (and protocol) of that page. 

If people want startURLs + their subdomains to be crawled, or start URLs + allowing URLs in a page defined as other startURLs to be crawled... they would use regular filtering for that.  In other words, we do not need to get too fancy with that feature.  

I am hoping this is along the line of what you were after, as I am marking this as a feature request. :-)

**Comment by csaezl** [comment]
That is it!. Thank you for taking the time to understand my need. 
Will you implement this feature ia a near future?.

**Comment by essiembre** [comment]
I wish, but I can't tell when I'll get to it given my busy schedule these days.  What I can tell you is this does not look like a huge level of effort.  That's the most positive answer I can give you right now. :-)

**Comment by csaezl** [comment]
Thank you

**Comment by essiembre** [comment]
Since I marked it as a feature request, let's keep it open until implemented.

**Comment by essiembre** [comment]
I figured a simple way to provide what you want.  I made a new [2.3.0-SNAPSHOT](http://www.norconex.com/collectors/collector-http/download) release that includes a new flag on the link extractor to ignore "external" URLS.  In its simplest form (assuming defaults), the config goes like this:

```xml
<linkExtractors>
  <extractor ignoreExternalLinks="true" />
</linkExtractors>
```
With this new flag, you should no longer need to user filters to achieve the same effect.  Let me know if how it goes.

**Comment by csaezl** [comment]
It works fine. Does it process subdomains?

**Comment by essiembre** [comment]
When that feature is enabled, no, it does not. For this, you would have to use regular filters instead, or define extra start URLs for the sub-domains you want.

**Comment by csaezl** [comment]
Trying to accomplish what is derived from this issue title "Several URLs processed one by one", I run a collector with 1 thread, several URL and ```<extractor ignoreExternalLinks="true" />```. The URLs were no processed one after one but mixed. Is there a way o processing them one after one, so in the log I have all the information together related to each URL?.

**Comment by essiembre** [comment]
What do you mean one after one?  With a single thread, there shall be only one URL processed at a time.  Do you have evidence of the opposite?  If so please submit a relevant portion of your log that shows the problem.

**Comment by essiembre** [comment]
Re-reading this.. ignore my previous comment.  It probably looks like they are mixed because of the redirect changing the URL on us.  But they should still be considered one by one like you want.  Make sense?

**Comment by csaezl** [comment]
Taking again the first example on this issue:
```
      <startURLs>
          <url>URL-A</url>      
          <url>URL-B</url>      
      </startURLs>
```
I need the crawler not begin proccessing URL-B, in any way, until the process on URL-A has finished. Even with several threads.

**Comment by essiembre** [comment]
Do you mean URL-A and all its descendants should be done before URL-B starts?  Or just that single page URL-A must be done before the single page URL-B starts?

If you are multi-threaded, there is no such guarantee for both scenarios.  They are run concurrently and anyone can finish first.  If the order had to be respected, they would need to be run sequentially, and that would not be multi-threading. :-).

If you are running single-threaded, the order these two URLs will be processed will be respected, but not their descendants.  If you truly want to isolate the crawling of each sites, you need to use different crawlers for each.  

But isn't your original concern addressed by the "Stay on site" feature, which is resolved by the ``<extractor ignoreExternalLinks="true" />``?

If you are talking about two different things, please create a separate ticket so we can close this one and we do not mix the two.  It is getting confusing otherwise.

**Comment by csaezl** [comment]
Of course, I mean the page and its descendants, not the single page. That was the original need, that is, to process first ```URL-A```, then ```URL-B```. not both at the same time, with all threads defined working on just one URL-x (and its descendants).

In the process of explaining my need, ```<extractor ignoreExternalLinks="true" />``` appeared solving something related with the issue, but not with the original need. By the way, this feature works fine for me.

So, it could work as I need if the collector doesn't inject all the ```<startURLs>``` at the beginning of the crawling process, but just the first one. When this first one is done (and its descendants), the crawler can inject the second one, and so on. 

This way of processing has to be taken as an option.

**Comment by essiembre** [comment]
So really what you want is to process one "site" at a time?  

It sounds good in principle, but raises many questions and issues that makes it not such an attractive idea to implement (and would take a while). I can get into it in more details if you wish, but instead for now I'll propose a solution that may give you what you want right away.

I understand your main objective is to have logs for each site together.  A mix of using grep and sort commands can help you do this, but you can do this via configuration. Since the configuration files are really Velocity templates, you can do something like this:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<httpcollector id="My Collector">
    ...
    <crawlers>
        #set( $allURLs = ["http://myurl1.com", "http://myurl2.com", "http://myurl3.com"] )
        #foreach( $oneURL $allURLs )
            <crawler id="crawler_${foreach.count}">
                ...
                    <startURLs><url><![CDATA[${oneURL}]]></url></startURLs>
                ...
            </crawler>
        #end
    </crawlers>
</httpcollector>
```

If having different logs is not enough for you and you really want them to run sequentially, I am afraid the only way for now would be to write a shell script that will create and run one collector instance for each URL, one after the other (same logic as above, but in a shell script). 

Can the above work for you?

**Comment by csaezl** [comment]
In my case I think the best way to implement it is via shell script. Thank you for your advice.

**Comment by essiembre** [comment]
You are welcome!  I'll close this issue since it is partly resolved by the ``ignoreExternalLinks`` (and hopefully simple scripting will do the rest).  

If there are no way to get what you want via scripting for "processing start URLs and their descendants sequentially", create a new issue just for that and give a use case or two that explain why such feature is needed and why it cannot be addressed with simple scripting.
